### DISCUSSION TITLE
Field Oriented Control

### DISCUSSION THREAD

**Discussion by IhorNehrutsa** [CAP1Sup/Intellistep#44] [General]
https://github.com/simplefoc/Arduino-FOC
https://docs.simplefoc.com/stepper_motors
https://docs.simplefoc.com/studio

**Comment by CAP1Sup** [comment]
Wouldn't this cause a huge performance impact? I could integrate it, it would just take a lot of time. Also, I think that we'd have to write the encoder file as a class instead of a collection of functions, as well as possibly some of the A4950 commands and outputs. It does look like an excellent option though.

**Reply by IhorNehrutsa** [reply]
This is for info and study.
I'm not sure if full integration is needed.
Very likely not.

But it will be great to integrate the **Commander** interface
https://docs.simplefoc.com/commander_interface
and use **SimpleFOCStudio**
https://docs.simplefoc.com/studio
https://github.com/JorgeMaker/SimpleFOCStudio
to tune and configure the **Intellistep** driver.

As stated in https://github.com/CAP1Sup/Intellistep#readme
```
Future Features:

- Manual tuning utility for manual PID tuning
```

It's not a request/demand for you, I can try to investigate/develop this functionality.

**Reply by IhorNehrutsa** [reply]
And I agree about the A4950. Set it onto the TODO list.

**Reply by CAP1Sup** [reply]
I'm a little concerned about the applicability of the SimpleFOC library. Sure, it's got good features. However, not only would we have to write custom classes for the motor driver and encoder, but we'd also have to write one to handle the motor in general using SimpleFOC commands. Also, SimpleFOC uses voltage control and a different stepping scheme, so I'm not confident that it's even possible to use the library without large modifications. Besides, the library will also probably have performance issues compared to how quick we've gotten the stepping routine now. I'm thinking it might be better to add FOC (non-library based) using our current setup, as well as fixing the minor stuff like the PID not working properly. I do like the idea of possibly piggybacking on their utility, that might be really nice to use. For right now, I want to put some time into the basic correction system as well as the PID later. It would be extremely helpful if you would be able to get a head start on finishing the PID. I was also thinking about the possibility of using the last dip to switch between PID and simple correction, as well as some simple G-codes, but I'm getting ahead of myself. I also wanted to write the encoder functions into a class and integrate the PID and encoder classes underneath the motor class. Let me know what you think. I really appreciate your help. It's a bit lonely with only myself working on the project

**Reply by xerootg** [reply]
Here ya go. Drivers written.

https://github.com/xerootg/btt-s42b-simplefoc

There's a step/dir interface too, just gotta call the constructor with the right pins and set angle each iteration.

There's a good parser library and an autotune GUI.

That said... I like what you've got going on here,  I just haven't had time outside of work to really take a good bite out. I'm not sure I want to see SimpleFOC in this project, but there is still something to the SimpleFOC project.
